File: src/2016/day02/python/solution.py
PAD = (("1", "2", "3"), ("4", "5", "6"), ("7", "8", "9"))
COMPLEX_PAD = (
    ("x", "x", "1", "x", "x"),
    ("x", "2", "3", "4", "x"),
    ("5", "6", "7", "8", "9"),
    ("x", "A", "B", "C", "x"),
    ("x", "x", "D", "x", "x"),
)

DIRECTION_MAP = {"U": (-1, 0), "R": (0, 1), "D": (1, 0), "L": (0, -1)}


def format_data(input: str) -> list[str]:
    return input.strip().splitlines()


def part1(input: str) -> str:
    instructions = format_data(input)
    y, x = 1, 1
    code = ""

    for instruction in instructions:
        for move in instruction:
            dy, dx = DIRECTION_MAP[move]
            y, x = y + dy, dx + x
            y, x = min(y, 2), min(x, 2)
            y, x = max(y, 0), max(x, 0)

        code = code + PAD[y][x]

    return code


def is_valid_complex_coordinate(y: int, x: int) -> bool:
    return 0 <= y < 5 and 0 <= x < 5 and COMPLEX_PAD[y][x] != "x"


def part2(input: str) -> str:
    instructions = format_data(input)
    y, x = 1, 1
    code = ""

    for instruction in instructions:
        for move in instruction:
            dy, dx = DIRECTION_MAP[move]
            ny, nx = y + dy, x + dx

            if is_valid_complex_coordinate(ny, nx):
                y, x = ny, nx

        code = code + COMPLEX_PAD[y][x]

    return code

File: src/2016/day02/typescript/solution.ts
const PAD = [
  ['1', '2', '3'],
  ['4', '5', '6'],
  ['7', '8', '9'],
];
const COMPLEX_PAD = [
  ['x', 'x', '1', 'x', 'x'],
  ['x', '2', '3', '4', 'x'],
  ['5', '6', '7', '8', '9'],
  ['x', 'A', 'B', 'C', 'x'],
  ['x', 'x', 'D', 'x', 'x'],
];

const DIRECTION_MAP = { U: [-1, 0], R: [0, 1], D: [1, 0], L: [0, -1] };

function formatData(input: string): string[] {
  return input.trim().split('\n');
}

function isValidSimple(y: number, x: number): boolean {
  return 0 <= y && y <= 2 && 0 <= x && x <= 2;
}

function isValidComplex(y: number, x: number): boolean {
  return 0 <= y && y <= 4 && 0 <= x && x <= 4 && COMPLEX_PAD[y][x] !== 'x';
}

function solve(
  input: string,
  pad: string[][],
  isValid: (y: number, x: number) => boolean,
  startY: number,
  startX: number,
): string {
  const instructions = formatData(input);
  let code = '',
    y = startY,
    x = startX;

  for (const instruction of instructions) {
    for (const move of instruction) {
      const [dY, dX] = DIRECTION_MAP[move];
      const nY = y + dY,
        nX = x + dX;
      if (isValid(nY, nX)) {
        y = nY;
        x = nX;
      }
    }
    code += pad[y][x];
  }

  return code;
}

export function part1(input: string): number | string | undefined {
  return solve(input, PAD, isValidSimple, 1, 1);
}

export function part2(input: string): number | string | undefined {
  return solve(input, COMPLEX_PAD, isValidComplex, 2, 0);
}
